### ISSUE TITLE
Branch out older version prior to development of a new one

### ISSUE THREAD

**Issue by aprokop** [trilinos/ForTrilinos#2]
- [ ] Create v1 version branch
- [ ] Make clear in documentation and on front page README that the old v1 version is unstable and the master branch is for development
- [x] Document the current testing status
- [x] Make sure the test pass so that we are able to enable CI

**Comment by bmpersc** [comment]
@aprokop @jwillenbring If this project is going to be resurrected we will need to have a plan on how to maintain compatibility with master, or at the very least for releases. Currently I believe all the branches on this repo since it was broken out point to the same commit so there has been no real concern about what has happened to it.

I don't know what the plans and expectations for this renewed effort so I can't suggest something at this point.

**Comment by aprokop** [comment]
@bmpersc ECP proposal for ForTrilinos was accepted. The work on a new version of ForTrilinos is to begin soon. My original thought was to preserve the existing code by putting it in a unmutable branch, and work on a newer version in the master. What compatibility issues did you have in mind? Also, do you know of any users of the existing code that could be affected by new developments?

**Comment by bmpersc** [comment]
@aprokop I guess the primary issue is that I misunderstood what the v1 branch was going to be. From the context of this  issue indicating new work was going to be done assumed that v1 was going to be where the work was to be done. From your last comment it is now clear that v1 will be the old code and new work will be on the develop branch, I hope.

As for compatibility, the concerns there was based on the previous faulty assumption that new work was going to be on v1 until "ready" and then merged into develop or master all at once. Since that doesn't seem like it will be the case then it isn't worth going into the possible issues with that model.

I don't know of anyone specifically using ForTrilinos. However, it is one of those sleeper codes. It was made for a specific purpose and presumably filled the role well. Since it isn't actively developed and is an interface to epetra which also isn't actively developed it is quite possible that someone is relying on it, but needs nothing new and as such never gets on our radar. I don't really know how to find people like that since the likely haven't made themselves known to us. Maybe @jwillenbring  would know of some people we could ask.  If there are such people either they are on a static version of Trilinos and we don't need to worry too much about breaking them. Or they only use our tarball releases, in which case we need to make sure to have the releases be stable. There is also a good chance no one really cares about ForTrilinos as it is and none of this is important.

**Comment by maherou** [comment]
Brent,

I don’t understand the concern you are raising.  Can you provide more detail?

Thanks.

Mike

**Comment by bmpersc** [comment]
@maherou There was a misunderstanding on my part, my previous comment explains my original thinking, but the concerns are moot now. I'd say I'm neutral to this at this point, however, there is a very slight concern on if anyone is using the package happily as it is and what will happen to them. But as I stated in my previous comment I think that is likely to be a very low probability.

**Comment by aprokop** [comment]
Depends on #4 as we want to have clean testing before branching out.

**Comment by aprokop** [comment]
Hmm, I did not realize that ForTrilinos maintains consistency with Trilinos, and still tags releases in the same way.
@bmpersc Is this practice to continue? Can we immediately break the backwards compatibility?
I'm OK with it, we would just need to record somewhere that the v1 API version corresponds to a specific Trilinos version (e.g., 12.10.1).

**Comment by bmpersc** [comment]
ForTrilinos is distributed as part of the Trilinos tarball so it will need to be tagged/branched with Trilinos for that. I didn't think about this issue specifically when I created the 12.10 branch and I didn't see anything countermanding that release model. Personally I'm okay with ForTrilinos changing it's release model since there is new work being done on ForTrilinos it might make sense not to release right now. As I mentioned above I don't know of anyone relying on ForTrilinos right now, but it is a sleeper code and we might not know who is using it until we break them.

Maybe we should have a chat with everyone involved, and at least @aprokop, @jwillenbring and others as warranted to decide what is the right branching/release model for ForTrilinos with respect to Trilinos at least in the short term (12.12, 12.14, etc.) Would you rather that be handled under a github issue or a phone conversation?

**Comment by jwillenbring** [comment]
I am fine with rethinking near-term ForTrilinos releases. I prioritize not interfering with the new effort much higher than anything related to the old version of the package.

**Comment by maherou** [comment]
I know of no one using the existing ForTrilinos capabilities.  We can tag a version of the old code and snapshot it into the Trilinos release, if we want to be particularly careful for the time being.  But in no way should the old version compromise our plans or progress on the new version.

**Comment by aprokop** [comment]
OK. I think the easiest thing would be to do the following:
- [ ] Create a new `v1` branch as described above
- [ ] Keep ForTrilinos being distributed as part of Trilinos tarball as @bmpersc described above, but in the near future create tags using the `v1` branch, and put that code in Trilinos tarball
- [ ] Start the development from scratch on the master, but not be include it in Trilinos tarballs until we decide it's ready in some fashion
- [ ] Document the process in README
- [ ] Switch to using master in tarballs

This way, we keep carrying the same old ForTrilinos code in Trilinos tarballs, but completely refactor the master branch.

**Comment by jwillenbring** [comment]
This sounds very reasonable. Would you prefer to switch directly from the old ForTrilinos to the new one at some point, or work toward removing the current version, then adding in the new one when it is ready?

**Comment by aprokop** [comment]
@jwillenbring I'm not sure at this point but I hope it's going to be a direct switch once we have a reasonable subset of functionality available in the new version. Carrying both at the same time would be troublesome as they may not sure the same approach towards interface.

**Comment by aprokop** [comment]
@jwillenbring One more thing: can we generate trilinos.org documentation/Doxygen for ForTrilinos using v1 branch?

**Comment by jwillenbring** [comment]
ForTrilinos V1 has been essentailly static. We have not been updating the Doxygen since the github move.

**Comment by aprokop** [comment]
That works, lets keep it this way until we are ready. Maybe just update the package page to describe the split.

**Comment by bmpersc** [comment]
I like what is being suggested here for dealing with ForTrilinos and branching Trilinos, but I do have some concerns.

The biggest concern I have about this is what this would mean for people who want to get Trilinos branches through github and not our tarballs. If they need some of the external repos then there is a discontinuity on how to get the appropriate sha1 from each repo. Some will be checkout the corresponding branch where ForTrilinos will be only tagged on a stale branch.

This probably is a bigger issue of how Trilinos will handle branching and external repos. Currently the git-dist tool won't support the model suggested here and I don't believe it should given what its primary use case is. I think the activity here is emphasizing the need for Trilinos to better define the role of external repos in the Trilinos life cycle at least for releases if nothing else.

**Comment by aprokop** [comment]
@bmpersc I don't understand what you are saying.

> If they need some of the external repos then there is a discontinuity on how to get the appropriate sha1 from each repo.

Not sure what you mean. Could you provide a concrete example?

> Some will be checkout the corresponding branch where ForTrilinos will be only tagged on a stale branch.

Again, have trouble understanding.

**Comment by jwillenbring** [comment]
@aprokop @bmpersc brings up a good point. The existing tools don't support using different branches for different repos. So, if people are trying to get both the Trilinos and ForTrilinos repos, they will have no real way to pull or even know to pull a v1 branch.

There are at least 2 alternatives to consider to revise:

>start the development from scratch on the master, but not be include it in Trilinos tarballs until we decide it's ready in some fashion

1. Start the development from scratch on a develop branch, and don't update master until the new version becomes the official version. Until then, the current master will be included in Trilinos tar balls, but the development version can be tested nightly.

2. Use a new branch for the development work instead of a new branch for the old version.

Thoughts?

**Comment by bmpersc** [comment]
@jwillenbring @aprokop, I wasn't trying to suggest that what was previously proposed was flawed to a point where we should redo it. More I was trying to point out that the proposal only addressed the releases when they are distributed as a tarball. Since the Trilinos repo is so readily accessible it is likely that people will access releases directly through branches now. And if they do that and need external repos then making how you get the branch for external repo A and external repo B different will add to the complexity of using Trilinos which I'd argue is already pretty high.

Mostly I was hoping that pointing out this inconsistency and the fact our existing tools won't support the proposed model would lead to a discussion about how to properly support all external repos. Since ForTrilinos is the first and currently only actively developed external repo that is officially part of Trilinos this is the first time that we've had to really consider these issues. In this particular case there are some unique issues, like the fact that there is an old code base that is going to be completely rewritten and we need to address those. However, I believe that there is a lot that could be standardized for all external repos too.

**Comment by aprokop** [comment]
@bmpersc @jwillenbring 

So, do we all agree that if we go with Jim's suggestion, and swap the branches, everything is going to be fine and dandy? I'm fine with that. The new checklist would be smth like this:

- [ ] Create `develop` branch
- [ ] Update master to indicate that it is a stale branch, and that development is done in `develop`.
- [ ] Start the development from scratch on `develop`
- [ ] Merge `develop` into `master` when ready

I think this would not require any changes to the current process. Any issues with calling it `develop`, same as in Trilinos proper?

**Comment by aprokop** [comment]
OK, I swapped the branches. So `master` now is the old master with the only change is that in README.md we indicate that it corresponds to stale development. The new `develop` branch is for current development from scratch.

**Comment by aprokop** [comment]
I'm closing the ticket. Feel free to reopen if I missed something and some workflow is broken.

**Comment by aprokop** [comment]
I'm switching the default ForTrilinos branch to `develop`. @bmpersc @jwillenbring Please let me know if it breaks any of your Trilinos tarball releases, and I'll switch it back.
